xla/pjrt/pjrt_stream_executor_client.cc

```diff
@@ -3448,7 +3460,32 @@ PjRtStreamExecutorClient::Compile(mlir::ModuleOp module,
       module, xla_computation,
       /*use_tuple_args=*/options.parameter_is_tupled_arguments,
       /*return_tuple=*/false));
-  return Compile(xla_computation, options);
+
+  TF_ASSIGN_OR_RETURN(std::vector<LayoutMode> arg_layout_modes,
+                      GetArgLayoutModes(module));
+  TF_ASSIGN_OR_RETURN(std::vector<LayoutMode> out_layout_modes,
+                      GetOutputLayoutModes(module));
+  TF_ASSIGN_OR_RETURN(std::vector<MemorySpaceColor> arg_memory_spaces,
+                      GetArgMemoryKinds(module));
+  TF_ASSIGN_OR_RETURN(std::vector<MemorySpaceColor> out_memory_spaces,
+                      GetOutputMemoryKinds(module));
+
+  Shape result_layout;
+  ExecutableBuildOptions build_options = options.executable_build_options;
```

Comment:
If LayoutModesToXla always updates options.build_options.result_layout() in place, you do not need to pass result_layout to Compile, and read it from options instead. It does change the assumption of the newly added Compile method but it may be more consistent if you are using LayoutModesToXla. What do you think?

Reply:
The original version of Compile seems to go out of its way to keep the options intact (it takes a copy of the options early on and that unchanged copy is stored into the executable), so I was concerned that changing the options might have some bad consequences. If you think it's ok to mutate the options, I am happy to take that path.